[react-native] Upgrade to 0.64.2

Kudo opened the pull request:
# Why

Upgrade react-native to 0.64.2 for SDK 43

# How

- [x] Fork `react-native-lab/react-native` from `react-native` 0.64-stable branch and cherry-pick necessary patches from `exp-latest`
- [x] Bump package versions:
- react "16.13.1" → "17.0.1"
- react-dom "16.13.1" → "17.0.1"
- react-native "0.63.2" → "0.64.2"
- react-test-renderer "~16.13.1" → "~17.0.1"
- @types/react-native "~0.63.2" → "~0.64.12"
- @babel/core "^7.0.0" → "^7.12.9"
- metro-react-native-babel-preset "~0.59.0" → "~0.64.0"
- jest "^25.2.0" → "^26.6.3"
- [x] bare-expo: sync project setup from [upgrade helper](https://react-native-community.github.io/upgrade-helper/?from=0.63.2&to=0.64.2)
- [x] Fix JS Error parsing error from [upstream change](https://github.com/facebook/react-native/commit/9edfc43aad2f68be8a37)  
- [x] Expo Go for iOS:
- Folly.podspec -> RCT-Folly.podspec for old SDKs 
- Upgrade reanimated vendoring script to support RN 0.64
- Update reanimated install binding with RCTJSIExecutorRuntimeInstaller
- [x] Expo Go for Android:
- Fix hanging from loading bundle which ReactRootView not attached to ReactInstanceManager from [upstream change](https://github.com/facebook/react-native/commit/2c896d35782cd04c873aefadc947447cc30a7f60)


Could use this command to make review easier:
`git diff master...HEAD -- . ':!android/ReactCommon' ':!android/ReactAndroid/src'`

# Test Plan

- Bare Expo
- [x] iOS runs
- [x] Android runs
- [x] Web runs 
- Expo Go with NCL
- [x] unversioned QA on iOS
- [x] unversioned QA on Android
- [x] Could load SDK 42 project on iOS
- [x] Could load SDK 42 project on Android   
- Shell app
- [x] iOS build passed
- [x] Android build passed

Kudo committed:
upgrade react-native fork

Kudo committed:
Bump package versions

Kudo committed:
fix metro exclusionList

Kudo committed:
upgrade bare-expo/bare-sandbox android project

Kudo committed:
upgrade bare-expo/bare-sandbox ios project

Kudo committed:
Fix for RCTTurboModuleManagerDelegate change

Kudo committed:
update reanimated podspec transform for RN 0.64

Kudo committed:
adopt RCTJSIExecutorRuntimeInstaller: https://github.com/software-mansion/react-native-reanimated/commit/1d3cfd746c2

Kudo committed:
backport Folly.podspec as vendored/shared/RN063Folly.podspec

Kudo committed:
removed unused folly version in transform script

Kudo committed:
et update-rn

Kudo committed:
fix error from react-native-codegen

Kudo committed:
Use RCT-Folly.podspec for RN063 to keep thing simpler

Kudo committed:
[android] Fix expo go does not show new contents for loading bundle.

The ReactRootView does not attach to ReactInstanceManager for our call flow.
Keep waiting in `waitForReactRootViewToHaveChildrenAndRunCallback`

Upstream change from
https://github.com/facebook/react-native/commit/2c896d35782cd04c873aefadc947447cc30a7f60

Kudo committed:
Fix error from error reporting.

Upstream change: https://github.com/facebook/react-native/commit/9edfc43aad2f68be8a37c916779681cd9ad13fee

Kudo committed:
Add back androidx.appcompat bumping

https://github.com/expo/expo/commit/58fa52e3e3e29b7a852a814b2d5bb54f4a8e1065

Kudo committed:
react-native sync: [android] Upgrade androidx.appcompat to 1.2.0

Kudo commented:
after rebase, versioned expo go for ios broken at reanimated. still in check.

Kudo committed:
Backport Folly -> RCT-Folly for reanimated in old SDK

Kudo committed:
fix ci shell-app-android that should checkout react-native fork

Kudo committed:
Support building unversioned ios shell app

Kudo committed:
build shell with s3 upload should be versioned

Kudo commented:
shell app building failures are fixed.
could start to review 🚀

brentvatne commented:
android/ReactAndroid/build.gradle:

```diff
-    compileSdkVersion 30
+    compileSdkVersion 29
```

was this intentional? related: https://github.com/expo/expo/pull/11647

android/expoview/src/main/java/host/exp/exponent/experience/ReactNativeActivity.java:

```diff
+    // Requesting layout to make sure {@link ReactRootView} attached to {@link ReactInstanceManager}
+    mReactRootView.call("requestLayout");
+
```

it might be worth adding more to this comment - clarify that this was found in upgrading react native 63 to 64 and that if this code isn't present then the app will hang when it loads, or something like that :)

Kudo committed:
fix compileSdkVersion be downgraded after upgrade

Kudo committed:
add more comment for requestLayout

Kudo commented:
> android/ReactAndroid/build.gradle:
> 
> ```diff
> -    compileSdkVersion 30
> +    compileSdkVersion 29
> ```
> 
> was this intentional? related: #11647

good catch!
updated in this commit: https://github.com/expo/expo/pull/13793/commits/7355db8f11dce35eb149d23f9d9099b2370cbe06

> 
> android/expoview/src/main/java/host/exp/exponent/experience/ReactNativeActivity.java:
> 
> ```diff
> +    // Requesting layout to make sure {@link ReactRootView} attached to {@link ReactInstanceManager}
> +    mReactRootView.call("requestLayout");
> +
> ```
> 
> it might be worth adding more to this comment - clarify that this was found in upgrading react native 63 to 64 and that if this code isn't present then the app will hang when it loads, or something like that :)

updated in https://github.com/expo/expo/pull/13793/commits/62e09769e9e165c3ecd0bf6087e92697f495fe6f


and rebase master to fix the conflict.
thanks for your review

tsapeta commented on tools/src/vendoring/config/expoGoConfig.ts:
```diff
@@ -37,29 +37,19 @@ const config: VendoringTargetConfig = {
-            'RNVERSION=64',
-            'RNVERSION=63'
-          );
+          // podspec.compiler_flags = podspec.compiler_flags.replace('RNVERSION=64', 'RNVERSION=63');
+          // podspec.xcconfig.OTHER_CFLAGS = podspec.xcconfig.OTHER_CFLAGS.replace(
+          //   'RNVERSION=64',
+          //   'RNVERSION=63'
+          // );
```
Is it no longer needed?

Kudo replied:
just to keep the code here in case reanimated upgrade and fallback default RN version as `0.65.0`.
https://github.com/software-mansion/react-native-reanimated/blob/4cbd70b5731970d0da1e2b1629f90afce98ae26a/RNReanimated.podspec#L10

bbarthec reviewed (approved):
I think that you did a great job and as long the main pipelines and apps are running I'm fine with it, but definitely the next step is to test the runtime extensively. I wonder if we have some kind of list describing `what should be tested after RN upgrade`. If no I guess we might want to create one during this time 😉

Kudo commented on .github/workflows/shell-app-ios.yml:
```diff
@@ -6,6 +6,9 @@ on:
       upload:
         description: 'type "upload" to confirm upload to S3'
         required: false
+      versioned:
+        description: 'Create shell workspace from newest SDK versioned environment'
+        required: false
```
for ios shell app to create workspace, classic xdl will install required packages from latest sdk.
in our case, it will install `react-native@https://github.com/expo/react-native/archive/sdk-42.0.0.tar.gz`
that's why to add packagesToInstallWhenEjecting override support for classic xdl in this commit: https://github.com/expo/xdl/commit/265186257fb7679c92a4cf9523a311cfb7c78202

Kudo commented:
> I think that you did a great job and as long the main pipelines and apps are running I'm fine with it, but definitely the next step is to test the runtime extensively. I wonder if we have some kind of list describing `what should be tested after RN upgrade`. If no I guess we might want to create one during this time 😉

the current testing scope includes these. from your experience, do you think if there are any else items need to pay attention?

# Test Plan

- Bare Expo
- [x] iOS runs
- [x] Android runs
- [x] Web runs 
- Expo Go with NCL
- [x] unversioned QA on iOS
- [x] unversioned QA on Android
- [x] Could load SDK 42 project on iOS
- [x] Could load SDK 42 project on Android   
- Shell app
- [x] iOS build passed
- [x] Android build passed

bbarthec commented:
@kudo, that list is good! I believe anything else that can came up later is too specific to test it beforehand 💪

Kudo commented:
thanks guys 🥰.
will rebase master and merge in the morning later (UTC+8). let's 🚀
